src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Text;
+
+namespace System.Net.NetworkInformation
+{
+    public class PhysicalAddress
+    {
+        private byte[] _address = null;
+        private bool _changed = true;
+        private int _hash = 0;
+
+        public static readonly PhysicalAddress None = new PhysicalAddress(Array.Empty<byte>());
+
+        public PhysicalAddress(byte[] address)
+        {
+            _address = address;
+        }
+
+        public override int GetHashCode()
+        {
+            if (_changed)
+            {
+                _changed = false;
+                _hash = 0;
+
+                int i;
+                int size = _address.Length & ~3;
+
+                for (i = 0; i < size; i += 4)
+                {
+                    _hash ^= (int)_address[i]
+                            | ((int)_address[i + 1] << 8)
+                            | ((int)_address[i + 2] << 16)
+                            | ((int)_address[i + 3] << 24);
+                }
+
+                if ((_address.Length & 3) != 0)
+                {
+                    int remnant = 0;
+                    int shift = 0;
+
+                    for (; i < _address.Length; ++i)
+                    {
+                        remnant |= ((int)_address[i]) << shift;
+                        shift += 8;
+                    }
+
+                    _hash ^= remnant;
+                }
+            }
+
+            return _hash;
+        }
+
+        public override bool Equals(object comparand)
+        {
+            PhysicalAddress address = comparand as PhysicalAddress;
+            if (address == null)
+            {
+                return false;
+            }
+
+            if (_address.Length != address._address.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < address._address.Length; i++)
```

Comment:
I missed the fact the code cached the hash: you're right: if the cache is different we can skip full equality checks.